chore(fqdn-template): fqdn templating move to specific folder and update documentation (#5354)
kubernetes-sigs/external-dns: docs/sources/fqdn-templating.md

> # FQDN Templating Guide

Reviewer: It's not really a source, like the other ones.

Wdyt about put this file in "advanced topics" section of the doc ?

Author: Make sense. Created folder `advanced` and moved few things there.

## What is FQDN Templating?

**FQDN templating** is a feature that allows to dynamically construct Fully Qualified Domain Names (FQDNs) using a Go templating engine.
Instead of relying solely on annotations or static names, you can use metadata from Kubernetes objects—such as service names, namespaces, and labels—to generate DNS records programmatically and dynamically.

This is useful for:

- Creating consistent naming conventions across environments.
- Reducing boilerplate annotations.
- Supporting multi-tenant or dynamic environments.
- Migrating from one DNS scheme to another
- Supporting multiple variants, such as a regional one and then one that doesn't or similar.

## How It Works

> ExternalDNS introduces a flag: `--fqdn-template`, which defines a Go template for rendering the desired DNS names.

Reviewer: ```suggestion
ExternalDNS has a flag: `--fqdn-template`, which defines a Go template for rendering the desired DNS names.
```

The template uses the following data from the source object (e.g., a `Service` or `Ingress`):

| Field         | Description                                                       |
|:--------------|:------------------------------------------------------------------|
| `Name`        | Name of the object (e.g., service)                                |
| `Namespace`   | Namespace of the object                                           |
| `Labels`      | Map of labels applied to the object                               |
| `Annotations` | Map of annotations                                                |
| `TargetName`  | For `Service`, it's the service name; for `Ingress`, the hostname |
| `Endpoint`    | Contains more contextual endpoint info, such as IP/target         |
| `Controller`  | Controller type (optional)                                        |

## Supported Sources

<!-- TODO: generate from code -->

| Source                 | Description                                                     | FQDN Supported |
|:-----------------------|:----------------------------------------------------------------|:--------------:|

> | `service`              | Queries Kubernetes Service resources for endpoints.             |       ✅        |
> | `ingress`              | Queries Kubernetes Ingress resources for endpoints.             |       ✅        |
> | `node`                 | Queries Kubernetes Node resources for endpoints.                |       ✅        |
> | `pod`                  | Queries Kubernetes Pod resources for endpoints.                 |       ❌        |
> | `fake`                 | Uses a fake source for testing purposes.                        |       ❌        |
> | `connector`            | Queries a custom connector source for endpoints.                |       ❌        |
> | `gateway-httproute`    | Queries HTTPRoute resources from the Gateway API.               |       ✅        |
> | `gateway-grpcroute`    | Queries GRPCRoute resources from the Gateway API.               |       ✅        |
> | `gateway-tlsroute`     | Queries TLSRoute resources from the Gateway API.                |       ❌        |
> | `gateway-tcproute`     | Queries TCPRoute resources from the Gateway API.                |       ✅        |
> | `gateway-udproute`     | Queries UDPRoute resources from the Gateway API.                |       ❌        |
> | `istio-gateway`        | Queries Istio Gateway resources for endpoints.                  |       ✅        |
> | `istio-virtualservice` | Queries Istio VirtualService resources for endpoints.           |       ✅        |
> | `cloudfoundry`         | Queries Cloud Foundry resources for endpoints.                  |       ❌        |
> | `contour-httpproxy`    | Queries Contour HTTPProxy resources for endpoints.              |       ✅        |
> | `gloo-proxy`           | Queries Gloo Proxy resources for endpoints.                     |       ❌        |
> | `crd`                  | Queries Custom Resource Definitions (CRDs) for endpoints.       |       ❌        |
> | `empty`                | Uses an empty source, typically for testing or no-op scenarios. |       ❌        |
> | `skipper-routegroup`   | Queries Skipper RouteGroup resources for endpoints.             |       ✅        |
> | `openshift-route`      | Queries OpenShift Route resources for endpoints.                |       ✅        |
> | `ambassador-host`      | Queries Ambassador Host resources for endpoints.                |       ❌        |
> | `kong-tcpingress`      | Queries Kong TCPIngress resources for endpoints.                |       ❌        |
> | `f5-virtualserver`     | Queries F5 VirtualServer resources for endpoints.               |       ❌        |
> | `f5-transportserver`   | Queries F5 TransportServer resources for endpoints.             |       ❌        |
> | `traefik-proxy`        | Queries Traefik Proxy resources for endpoints.                  |       ❌        |

Reviewer: ```suggestion
| `ambassador-host`      | Queries Ambassador Host resources for endpoints.                |       ❌        |
| `cloudfoundry`         | Queries Cloud Foundry resources for endpoints.                  |       ❌        |
| `connector`            | Queries a custom connector source for endpoints.                |       ❌        |
| `contour-httpproxy`    | Queries Contour HTTPProxy resources for endpoints.              |       ✅        |
| `crd`                  | Queries Custom Resource Definitions (CRDs) for endpoints.       |       ❌        |
| `empty`                | Uses an empty source, typically for testing or no-op scenarios. |       ❌        |
| `f5-transportserver`   | Queries F5 TransportServer resources for endpoints.             |       ❌        |
| `f5-virtualserver`     | Queries F5 VirtualServer resources for endpoints.               |       ❌        |
| `fake`                 | Uses a fake source for testing purposes.                        |       ❌        |
| `gateway-grpcroute`    | Queries GRPCRoute resources from the Gateway API.               |       ✅        |
| `gateway-httproute`    | Queries HTTPRoute resources from the Gateway API.               |       ✅        |
| `gateway-tcproute`     | Queries TCPRoute resources from the Gateway API.                |       ✅        |
| `gateway-tlsroute`     | Queries TLSRoute resources from the Gateway API.                |       ❌        |
| `gateway-udproute`     | Queries UDPRoute resources from the Gateway API.                |       ❌        |
| `gloo-proxy`           | Queries Gloo Proxy resources for endpoints.                     |       ❌        |
| `ingress`              | Queries Kubernetes Ingress resources for endpoints.             |       ✅        |
| `istio-gateway`        | Queries Istio Gateway resources for endpoints.                  |       ✅        |
| `istio-virtualservice` | Queries Istio VirtualService resources for endpoints.           |       ✅        |
| `kong-tcpingress`      | Queries Kong TCPIngress resources for endpoints.                |       ❌        |
| `node`                 | Queries Kubernetes Node resources for endpoints.                |       ✅        |
| `openshift-route`      | Queries OpenShift Route resources for endpoints.                |       ✅        |
| `pod`                  | Queries Kubernetes Pod resources for endpoints.                 |       ❌        |
| `service`              | Queries Kubernetes Service resources for endpoints.             |       ✅        |
| `skipper-routegroup`   | Queries Skipper RouteGroup resources for endpoints.             |       ✅        |
| `traefik-proxy`        | Queries Traefik Proxy resources for endpoints.                  |       ❌        |
```

```

This helps automate DNS record migration while maintaining service continuity.

### Multi-Variant Domain Support

You can also support regional variants or multi-tenant architectures, where the same service is deployed to different regions or environments:

```yaml
--fqdn-template='{{ .Name }}.{{ .Environment }}.{{.Region}}.example.com, {{ if eq .Environment "prod" }}{{ .Name }}.my-company.tld{{ end }}'
```

With additional context (e.g., annotations), this can produce FQDNs like:

```yml
api.dev.us-east-1.example.com
api.my-company.tld
```

This is helpful in scenarios such as:

- Blue/green deployments across domains
- Staging vs. production resolution
- Multi-cloud or multi-region failover strategies

## Tips

- If `--fqdn-template` is specified, ExternalDNS ignores any `external-dns.alpha.kubernetes.io/hostname` annotations.
- You must still ensure the resulting FQDN is valid and unique.
- Since Go templates can be error-prone, test your template with simple examples before deploying. Mismatched field names or nil values (e.g., missing labels) will result in errors or skipped entries.

## FaQ

### Can I specify multiple global FQDN templates?

> No, you can. Pass in a comma separated list to --fqdn-template. Beaware this will double (triple, etc) the amount of DNS entries based on how many services, ingresses and so on you have and will get you faster towards the API request limit of your DNS provider.

Reviewer: ```suggestion
Yes, you can. Pass in a comma separated list to --fqdn-template. Beware this will double (triple, etc) the amount of DNS entries based on how many services, ingresses and so on you have and will get you faster towards the API request limit of your DNS provider.
```

### Where to find template syntax

- [Go template syntax](https://pkg.go.dev/text/template)
- [Go func builtins](https://github.com/golang/go/blob/master/src/text/template/funcs.go#L39-L63)

### FQDN Templating, Helm and improper templating syntax

The user encountered errors due to improper templating syntax:

```yml
extraArgs:
  - --fqdn-template={{name}}.uat.example.com
```

The correct syntax should include a dot prefix: `{{.Name}}`.
Additionally, when using Helm's `tpl` function, it's necessary to escape the braces to prevent premature evaluation:

```yml
extraArgs:
  - --fqdn-template={{ `{{.Name}}.uat.example.com` }}
```

### Handling Subdomain-Only Hostnames

In [Issue #1872](https://github.com/kubernetes-sigs/external-dns/issues/1872), it was observed that ExternalDNS ignores the `--fqdn-template` when the ingress host field is set to a subdomain (e.g., foo) without a full domain.
The expectation was that the template would still apply, generating entries like `foo.bar.example.com.`
This highlights a limitation to be aware of when designing FQDN templates.

> > This currently not supported!!! Would expect external-dns to generate a dns record according to the fqdnTemplate

Reviewer: ```suggestion
> :warning: This is currently not supported ! User would expect external-dns to generate a dns record according to the fqdnTemplate
```
